### PR TITLE
feat: fold builder method

### PR DESCRIPTION
**Pull request by ctian1**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,4 +1,4 @@
 RPC_1=
-CONSENSUS_RPC_1=
-BEACONCHAIN_API_KEY_1=
-BEACONCHAIN_API_URL_1=https://beaconcha.in
\ No newline at end of file
+CONSENSUS_RPC_URL=
+BEACONCHAIN_API_KEY=
+BEACONCHAIN_API_URL=https://beaconcha.in
```
Comment (puma314) [inline .env.example:4]: Should we remove this since I just got rid of our subscription lol

File: Cargo.toml [hunk 2]
```diff
@@ -10,3 +10,6 @@ incremental = true
 
 [profile.bench]
 opt-level = 3
+
+[patch.'https://github.com/mir-protocol/plonky2.git']
+plonky2 = { git = "https://github.com/succinctlabs/plonky2.git", branch = "chris/recursion" }
```
Comment (puma314) [inline Cargo.toml:15]: nit: change this once plonky2 PR is merged

File: plonky2x/core/src/frontend/builder/io.rs [hunk 3]
```diff
@@ -147,16 +186,66 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
+                    verifier_data: None,
+                    closed: false,
+                })
+            }
+            CircuitIO::CyclicProof(_) => {}
+            _ => panic!("other io used already"),
+        };
+    }
+
+    pub fn close_cyclic_io(&mut self) {
```
Comment (puma314) [inline plonky2x/core/src/frontend/builder/io.rs:210]: can you add a comment here as to what is going on when you close the cyclic io?

File: plonky2x/core/src/frontend/builder/io.rs [hunk 4]
```diff
@@ -181,22 +270,39 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
         self.try_init_proof_io();
-        let proof = self.add_virtual_proof_with_pis(&child_circuit.data.common);
+        let proof = self.add_virtual_proof_with_pis(data);
         match self.io {
-            CircuitIO::RecursiveProofs(ref mut io) => io.input.push(proof.clone()),
+            CircuitIO::RecursiveProofs(ref mut io) => {
+                io.input.push(proof.clone());
+            }
+            CircuitIO::CyclicProof(ref mut io) => {
+                if io.proof.is_some() {
```
Comment (puma314) [inline plonky2x/core/src/frontend/builder/io.rs:282]: How would the proof already get set here?

File: plonky2x/core/src/frontend/builder/mod.rs [hunk 5]
```diff
@@ -87,13 +87,13 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
...
 
-        if let Ok(rpc_url) = env::var("CONSENSUS_RPC_1") {
+        if let Ok(rpc_url) = env::var("CONSENSUS_RPC_URL") {
             let client = BeaconClient::new(rpc_url);
             builder.set_beacon_client(client);
         }
 
-        if let Ok(api_url) = env::var("BEACONCHAIN_API_URL_1") {
-            if let Ok(api_key) = env::var("BEACONCHAIN_API_KEY_1") {
+        if let Ok(api_url) = env::var("BEACONCHAIN_API_URL") {
```
Comment (puma314) [inline plonky2x/core/src/frontend/builder/mod.rs:95]: do we use this?

File: plonky2x/core/src/frontend/builder/mod.rs [hunk 6]
```diff
@@ -259,6 +264,20 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
         }
     }
 
+    pub fn try_build(mut self) -> (CircuitBuild<L, D>, bool) {
```
Comment (puma314) [inline plonky2x/core/src/frontend/builder/mod.rs:267]: can you add a comment for what this is doing and why it's useful?

File: plonky2x/core/src/frontend/fold/generator.rs [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    <<L as PlonkParameters<D>>::Config as GenericConfig<D>>::Hasher: AlgebraicHasher<L::Field>,
+{
+    let mut input = circuit.input();
+    input.write::<Ctx>(ctx);
+    input.write::<Element>(element);
+    input.write::<Accumulator>(acc);
+    input.write::<Accumulator>(initial);
+    input.write::<U32Variable>(index);
+    input.data_write(circuit.data.verifier_data());
+    let proof = if let Some((proof, _)) = prev_result.take() {
```
Comment (puma314) [inline plonky2x/core/src/frontend/fold/generator.rs:113]: pls some comments

File: plonky2x/core/src/frontend/fold/mod.rs [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    let initial = builder.read::<Accumulator>();
+    let index = builder.read::<U32Variable>();
+
+    /*
+    Proof verifies:
+    (ctx, index, element, prev_acc, initial_acc) -> next_acc
+    if index === 0, disregard prev_acc and use empty initial
+    inner ctx == ctx
+    inner index = index - 1
+    inner elment: anything
```
Comment (puma314) [inline plonky2x/core/src/frontend/fold/mod.rs:149]: nit: typo

File: plonky2x/core/src/backend/circuit/input.rs [hunk 9]
```diff
@@ -15,6 +16,11 @@ pub enum PublicInput<L: PlonkParameters<D>, const D: usize> {
     Elements(Vec<L::Field>),
     RecursiveProofs(Vec<ProofWithPublicInputs<L::Field, L::Config, D>>),
     RemoteRecursiveProofs(Vec<ProofId>),
+    CyclicProof(
+        Vec<L::Field>,
+        Box<Option<ProofWithPublicInputs<L::Field, L::Config, D>>>,
```
[the remark this answers is not in the document]
Comment (ctian1) [reply plonky2x/core/src/backend/circuit/input.rs:21]: clippy was giving warnings that this variant of the enum is much bigger than the others

File: .env.example (hunk 1 above)
Comment (ctian1) [reply .env.example:4]: yeah we can

**Review by puma314 — APPROVED** [review]

**Pull request merged**